tests/001-handshake.cpp

```diff
@@ -108,68 +109,116 @@ namespace oxen::quic::test
             test_net.set_shutdown_immediate();
 
             opt::local_addr default_addr{}, local_addr{};
-            auto local_tls = GNUTLSCreds::make("./serverkey.pem"s, "./servercert.pem"s, "./clientcert.pem"s);
+
+            auto [local_tls, _] = defaults::tls_creds_from_ed_keys();
 
             auto ep_notls = test_net.endpoint(default_addr);
             auto ep_tls = test_net.endpoint(local_addr);
 
-            // ep_notls->listen();  // Shouldn't compile if uncommented!
-            // ep_notls->listen(local_tls, local_tls);  // Nor this
             REQUIRE_NOTHROW(ep_tls->listen(local_tls));
         };
+    };
 
-        SECTION("Endpoint::listen() + Endpoint::Connect() - Default addressing")
-        {
-            Network test_net{};
-            opt::local_addr default_addr{};
-            auto local_tls = GNUTLSCreds::make("./serverkey.pem"s, "./servercert.pem"s, "./clientcert.pem"s);
+    TEST_CASE("001 - Handshaking: Client Validation", "[001][client]")
+    {
+        auto client_established = callback_waiter{[](connection_interface&) {}};
+        auto server_established = callback_waiter{[](connection_interface&) {}};
 
-            auto ep = test_net.endpoint(default_addr);
+        Network test_net{};
 
-            REQUIRE_NOTHROW(ep->listen(local_tls));
-            REQUIRE_THROWS(ep->connect(default_addr, local_tls));
-        };
+        auto [client_tls, server_tls] = defaults::tls_creds_from_ed_keys();
 
-        SECTION("Endpoint::connect() - Immediate network shutdown")
-        {
-            Network test_net{};
-            test_net.set_shutdown_immediate();
-            auto server_tls = GNUTLSCreds::make("./serverkey.pem"s, "./servercert.pem"s, "./clientcert.pem"s);
-            auto client_tls = GNUTLSCreds::make("./clientkey.pem"s, "./clientcert.pem"s, "./servercert.pem"s);
+        opt::local_addr server_local{};
+        opt::local_addr client_local{};
 
-            opt::local_addr server_local{};
-            opt::local_addr client_local{};
+        auto server_endpoint = test_net.endpoint(server_local, server_established);
+        REQUIRE(server_endpoint->listen(server_tls));
 
-            auto server_endpoint = test_net.endpoint(server_local);
-            REQUIRE(server_endpoint->listen(server_tls));
+        SECTION("Endpoint::listen() + Endpoint::Connect() - Incorrect pubkey in remote")
+        {
+            uint64_t client_error{0};
 
-            opt::local_addr client_remote{"127.0.0.1"s, server_endpoint->local().port()};
+            auto client_closed = callback_waiter{[&client_error](connection_interface&, uint64_t) { client_error = 1000; }};
```

Comment:
Is there a particularly errcode value (passed in the second argument) for this mismatched-key failure?  Or to put it another way, how could an application distinguish between a bad pubkey, and, say, a timeout?

Reply:
A normal (not error-caused) closure would fire with an error code of 0, so any non-zero error code is a problem.

Currently, a pubkey that could not be validated is still an allowed connection, so it wouldn't fire the connection closed callback. Rather, it would fire the connection established callback, and the user checks `connection_interface.is_validated()`